services/autoscaling/tests/unit/conftest.py

```diff
@@ -40,12 +75,385 @@ def app_environment(
+    @retry(
+        retry=retry_if_exception_type(AssertionError),
+        reraise=True,
+        wait=wait_fixed(0.5),
+        stop=stop_after_delay(10),
+    )
+    async def _check_service_task_gone(service: Mapping[str, Any]) -> None:
+        print(
+            f"--> checking if service {service['ID']}:{service['Spec']['Name']} is really gone..."
+        )
+        assert not await async_docker_client.containers.list(
+            filters={
+                "label": [f"com.docker.swarm.service.id={service['ID']}"],
+            }
+        )
+        print(f"<-- service {service['ID']}:{service['Spec']['Name']} is gone.")
+
+    await asyncio.gather(*(_check_service_task_gone(s) for s in created_services))
+    await asyncio.sleep(0)
+
+
+@pytest.fixture
+def assert_for_service_state() -> Callable[
+    [aiodocker.Docker, Mapping[str, Any], list[str]], Awaitable[None]
+]:
+    async def _runner(
+        async_docker_client: aiodocker.Docker,
+        created_service: Mapping[str, Any],
+        expected_states: list[str],
+    ) -> None:
+        SUCCESS_STABLE_TIME_S: Final[float] = 3
+        WAIT_TIME: Final[float] = 0.5
+        number_of_success = 0
+        async for attempt in AsyncRetrying(
+            retry=retry_if_exception_type(AssertionError),
+            reraise=True,
+            wait=wait_fixed(WAIT_TIME),
+            stop=stop_after_delay(10 * SUCCESS_STABLE_TIME_S),
+        ):
+            with attempt:
+                print(
+                    f"--> waiting for service {created_service['ID']} to become {expected_states}..."
+                )
+                services = await async_docker_client.services.list(
+                    filters={"id": created_service["ID"]}
+                )
+                assert services, f"no service with {created_service['ID']}!"
+                assert len(services) == 1
+                found_service = services[0]
+
+                tasks = await async_docker_client.tasks.list(
+                    filters={"service": found_service["Spec"]["Name"]}
+                )
+                assert tasks, f"no tasks available for {found_service['Spec']['Name']}"
+                assert len(tasks) == 1
+                service_task = tasks[0]
+                assert (
+                    service_task["Status"]["State"] in expected_states
+                ), f"service {found_service['Spec']['Name']}'s task is {service_task['Status']['State']}"
+                print(
+                    f"<-- service {found_service['Spec']['Name']} is now {service_task['Status']['State']} {'.'*number_of_success}"
+                )
+                number_of_success += 1
+                assert (number_of_success * WAIT_TIME) >= SUCCESS_STABLE_TIME_S
+                print(
+                    f"<-- service {found_service['Spec']['Name']} is now {service_task['Status']['State']} after {SUCCESS_STABLE_TIME_S} seconds"
+                )
+
+    return _runner
+
+
+@pytest.fixture(scope="module")
```

Comment:
Since it was agreed upon not to use module scope in fixtures, could you add a comment with a reason why this is scoped module?
Could it be removed?

Reply:
so I do not think we agreed to not use them at all. We agreed not to use them in problematic cases:
- defining ENV variables by overriding a function fixture into a module fixture (such as the monkeypatch one)

Here, this fixture is about creating a mock threaded server of the AWS services (which by the way is also used in the dynamic-sidecar tests, storage and the api-server). Would you *really* want to have a server start and stop with each test? that would make the whole testing quite slow. Same we do not start and stop all the docker containers...

Reply:
OK my understanding of the problem was not complete. Thanks for clarifying this out.